GVFS/GVFS.Service/GvfsService.cs

```diff
@@ -230,72 +230,22 @@ private void HandleRequest(ITracer tracer, string request, NamedPipeServer.Conne
 
             using (ITracer activity = this.tracer.StartActivity(message.Header, EventLevel.Informational, new EventMetadata { { "request", request } }))
             {
-                switch (message.Header)
+                if (message.Header == NamedPipeMessages.EnableAndAttachProjFSRequest.Header)

```

Comment:
Here is how I thought about this
```
public class RequestProcessor
{
    public void HandleRequest(ITracer tracer, string request, NamedPipeServer.Connection connection)
    {
        <Code to setup message that is above this comment>
        this.HandleMessage(...)
    }

    protected virtual void HandleMessage(
            ITracer tracer,
            string etwArea,
            RepoRegistry repoRegistry,
            NamedPipeMessages.Message message,
            NamedPipeServer.Connection connection)
    {
        <Do common stuff>
    }
}

public class WindowsRequestProcessor
{
    protected override void HandleMessage(
            ITracer tracer,
            string etwArea,
            RepoRegistry repoRegistry,
            NamedPipeMessages.Message message,
            NamedPipeServer.Connection connection)
    {
        <do windows work>
        // if message wasn't processed
        base.HandleMessage(tracer, etwArea, repoRegistry, message, connection)
    }
}
```
Then in the constructor you would create the class 
`this.requestProcessor = new (WindowsRequestProcessor/RequestProcessor)`
and then set the handler of the named pipe
`using (NamedPipeServer pipeServer = NamedPipeServer.StartNewServer(pipeName, this.tracer, this.requestProcessor.HandleRequest))`

Not sure how much better this is but it makes more sense to me and abstracts things a little better.  If we could find a way to remove the `NamedPipeServer.Connection connection` we could then easily write some unit tests.

Reply:
that looks a lot cleaner. updated.